sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -713,27 +713,22 @@ object HiveTypeCoercion {
       @Nullable val ret: Expression = (inType, expectedType) match {
 
         // If the expected type is already a parent of the input type, no need to cast.
-        case _ if expectedType.isSameType(inType) => e
+        case _ if expectedType.acceptsType(inType) => e
```

Comment:
Why we need `isSameType`? I think the logic should be:

1. first check whether the input type is acceptable(i.e. expectedType.acceptsType(inType) returns true). If it is, do nothing here.
2. If input type is not acceptable, follow special rules below.

It looks to me we only need `acceptsType`.